gnovm/pkg/gnolang/type_check.go

```diff
@@ -997,6 +998,40 @@ func assertValidAssignLhs(store Store, last BlockNode, lx Expr) {
 	}
 }
 
+func assertValidAssignRhs(store Store, last BlockNode, n Node) {
+	var exps []Expr
+	switch x := n.(type) {
+	case *ValueDecl:
+		exps = x.Values
+	case *AssignStmt:
+		exps = x.Rhs
+	default:
+		panic(fmt.Sprintf("unexpected node type %T", n))
+	}
+
+	for _, exp := range exps {
+		tt := evalStaticTypeOfRaw(store, last, exp)
+		if tt == nil {
+			switch x := n.(type) {
+			case *ValueDecl:
+				if x.Type != nil {
+					continue
```

Comment:
this is not correct:
```go
package main

func main() {
	var t int = nil
}
```

Reply:
The assignment is valid, but the name is not defined. For instance:

```
package main

func main() {
	null := 1
	var i int
	i = null
	print("here", i)
}
```

This case is already addressed, but the error message could be improved.

The situation is similar in this example:

```
package main

func main() {
	var i int = a
	print("here", i)
}
```

Reply:
https://go.dev/play/p/3zb0tttu0It

Reply:
My bad I use a wrong example. I will have a look

Reply:
Should we address this case here?  (This expression is valid, but the types are different.)

Perhaps we should manage it within the same branch as the example that I write after? 

```go
package main

func main()  {
  b := 1
  var i int = &b
}
```

Reply:
this is a valid RHS expression, but it is not type-compatible for assignment. As a result, it should trigger a 'cannot use xxx as xxx' error.  I think it's ok to put it in another fix.

Reply:
I think we should treat both as type-compatible errors for easier handling. In Go, I think both examples display the type-incompatible error:
```
package main

func main() {
	b := 1
	var i int = &b
}
```
Error: cannot use &b (value of type *int) as int value in variable declaration

```
package main

func main() {
	var i int = nil
}
```

Error: cannot use nil as int value in variable declaration

Reply:
see https://github.com/gnolang/gno/pull/3068